<a href="https://colab.research.google.com/github/1991Ab/Capstone-project-2-Taxi-trip-time-Prediction/blob/main/Taxi_trip_time_Prediction_individual_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
a= input()
b= input()
print(a)
print(b)

a
b
a
b


# **Project Name**    -  NYC Taxi Trip Time Prediction



##### **Project Type**    - Regression
##### **Contribution**    - Individual


# **Problem Statement**


**BUSINESS PROBLEM OVERVIEW**


Your task is to build a model that predicts the total ride duration of taxi trips in New York City. Your primary dataset is one released by the NYC Taxi and Limousine Commission, which includes pickup time, geo-coordinates, number of passengers, and several other variables.

1.   Well-structured, formatted, and commented code is required. 
2.   Exception Handling, Production Grade Code & Deployment Ready Code will be a plus. Those students will be awarded some additional credits. 
     
     The additional credits will have advantages over other students during Star Student selection.
       
             [ Note: - Deployment Ready Code is defined as, the whole .ipynb notebook should be executable in one go
                       without a single error logged. ]

3.   Each and every logic should have proper comments.
4. You may add as many number of charts you want. Make Sure for each and every chart the following format should be answered.
        

```
# Chart visualization code
```
            

*   Why did you pick the specific chart?
*   What is/are the insight(s) found from the chart?
* Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

5. You have to create at least 15 logical & meaningful charts having important insights.


[ Hints : - Do the Vizualization in  a structured way while following "UBM" Rule. 

U - Univariate Analysis,

B - Bivariate Analysis (Numerical - Categorical, Numerical - Numerical, Categorical - Categorical)

M - Multivariate Analysis
 ]





6. You may add more ml algorithms for model creation. Make sure for each and every algorithm, the following format should be answered.


*   Explain the ML Model used and it's performance using Evaluation metric Score Chart.


*   Cross- Validation & Hyperparameter Tuning

*   Have you seen any improvement? Note down the improvement with updates Evaluation metric Score Chart.

*   Explain each evaluation metric's indication towards business and the business impact pf the ML model used.




















# ***Let's Begin !***

## ***1. Know Your Data***

### Import Libraries

In [2]:
# Import Libraries
import numpy as np
import pandas as pd
from numpy import math
from numpy import loadtxt
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from matplotlib import rcParams
!pip install pymysql
import pymysql
from sqlalchemy import create_engine
from sqlalchemy.pool import NullPool

import numpy as np
import seaborn as sns
from scipy.stats import *
import math

from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn import metrics
from sklearn.metrics import roc_curve
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RepeatedStratifiedKFold
from xgboost import XGBClassifier
from xgboost import XGBRFClassifier
from sklearn.tree import export_graphviz

!pip install shap==0.40.0
import shap 
import graphviz
sns.set_style('darkgrid') 

import warnings
warnings.filterwarnings('ignore')

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 43 kB 977 kB/s 
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 571 kB 5.0 MB/s 


### Dataset Loading

In [3]:
# Load Dataset
# Mounting the Google Drive
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [4]:
# Importing the dataset
file_path = "/content/drive/MyDrive/Capstone project-2/NYC Taxi Data (1).csv"
df = pd.read_csv(file_path)


### Dataset First View

In [5]:
# Dataset First 
df.head()

id  vendor_id      pickup_datetime     dropoff_datetime  \
0  id2875421          2  2016-03-14 17:24:55  2016-03-14 17:32:30   
1  id2377394          1  2016-06-12 00:43:35  2016-06-12 00:54:38   
2  id3858529          2  2016-01-19 11:35:24  2016-01-19 12:10:48   
3  id3504673          2  2016-04-06 19:32:31  2016-04-06 19:39:40   
4  id2181028          2  2016-03-26 13:30:55  2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude store_and_fwd_flag  trip_duration  
0         40.765602                  N            455  
1         40.731152                  N            663  
2         40.710087                  N           2124  
3         40.706718                  N            429  
4         40.782520                  N            435

### Dataset Rows & Columns count

In [6]:
# Dataset Rows & Columns 
df.shape

(1458644, 11)

We can see that the NYC taxi trip time prediction data set contains 1458644 rows and 11 columns

### Dataset Information

In [7]:
# Dataset Info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458644 entries, 0 to 1458643
Data columns (total 11 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   id                  1458644 non-null  object 
 1   vendor_id           1458644 non-null  int64  
 2   pickup_datetime     1458644 non-null  object 
 3   dropoff_datetime    1458644 non-null  object 
 4   passenger_count     1458644 non-null  int64  
 5   pickup_longitude    1458644 non-null  float64
 6   pickup_latitude     1458644 non-null  float64
 7   dropoff_longitude   1458644 non-null  float64
 8   dropoff_latitude    1458644 non-null  float64
 9   store_and_fwd_flag  1458644 non-null  object 
 10  trip_duration       1458644 non-null  int64  
dtypes: float64(4), int64(3), object(4)
memory usage: 122.4+ MB


Here we see the details of each column present in the dataset such as  their data type and count of values 

#### Duplicate Values

In [8]:
# Dataset Duplicate Value Count
len(df[df.duplicated()])

0

It is very essential to check if our data set has any duplicate values and remove them if any.We can find out the duplicate values using the function dataset.duplicated().  len(dataset[dataset.duplicated()]) gives the count of duplicate values.From the above code we can see that there are no duplicate values in the dataset.

#### Missing Values/Null Values

In [9]:
# Missing Values/Null Values Count
print(df.isnull().sum())

id                    0
vendor_id             0
pickup_datetime       0
dropoff_datetime      0
passenger_count       0
pickup_longitude      0
pickup_latitude       0
dropoff_longitude     0
dropoff_latitude      0
store_and_fwd_flag    0
trip_duration         0
dtype: int64


In [ ]:
# Visualizing the missing values
# Checking Null Value by plotting Heatmap
sns.heatmap(df.isnull(), cbar=True)

It is very important to check for missing values or NAN values as it has a very high impact on the model.Hence identifying the NAN values and treating them becomes very important.From the above data and graph we can see that the dataset has no missing or NAN values.

### What did you know about your dataset?

The dataset given is based on the  NYC Yellow Cab trip record data.We have to predict the taxi trip duration of various rides using various Machine Learning models.The data set has 'trip_duration' as a continous target variable and many independent variables like 'passenger_count' , 'pickup_datetime' , 'dropoff_datetime' etc.The  dataset has 1458644 rows and 11 columns.It has no NAN values and duplicate values.

## ***2. Understanding Your Variables***

In [ ]:
# Dataset Columns
df.columns

In [ ]:
# Dataset Describe
df.describe(include='all')

### Variables Description 

The following are the variables used in dataset

* #### id - a unique identifier for each trip
* #### vendor_id - a code indicating the provider associated with the trip record
* #### pickup_datetime - date and time when the meter was engaged
* #### dropoff_datetime - date and time when the meter was disengaged
* #### passenger_count - the number of passengers in the vehicle (driver entered value)
* #### pickup_longitude - the longitude where the meter was engaged
* #### pickup_latitude - the latitude where the meter was engaged
* #### dropoff_longitude - the longitude where the meter was disengaged
* #### dropoff_latitude - the latitude where the meter was disengaged
* #### store_and_fwd_flag - This flag indicates whether the trip record was held in vehicle memory before sending to the vendor because the vehicle did not have a connection to the server - Y=store and forward; N=not a store and forward trip
* #### trip_duration - duration of the trip in seconds

### Check Unique Values for each variable.

In [ ]:
# Check Unique Values for each variable.
for i in df.columns.tolist():
  print("No. of unique values in ",i,"is",df[i].nunique(),".")
#df.nunique()

## 3. ***Data Wrangling***

### Data Wrangling Code

In [ ]:
# Write your code to make your dataset analysis ready.
# Create a copy of the current dataset and assigning to df
new_df=df.copy()
# Checking Shape of True Value
new_df.shape


In [ ]:
#calculating the distance between pickup and dropoff locations using haversine formula

In [ ]:
!pip  install haversine
from numpy import math
from haversine import haversine
def trip_distance(df) :
  trip_pickup=(df['pickup_latitude'], df['pickup_longitude'])
  trip_dropoff=(df['dropoff_latitude'], df['dropoff_longitude'])
  return haversine(trip_pickup , trip_dropoff)

In [ ]:
#calculating the distance and creating a new column "distance" in the dataset
df['distance']=df.apply(lambda x:trip_distance(x) , axis=1)

In [ ]:
df['distance'].reset_index


In [ ]:
df

Calculating the customer with highest trip distance


In [ ]:
max_distance=df.groupby('id')['distance'].max().reset_index().sort_values(by='distance' , ascending=False)

In [ ]:
max_distance

In [ ]:
#Graphical Representation
sns.set_style('whitegrid')
plt.rcParams['font.size'] = 10
plt.rcParams['figure.figsize'] = (14,5)
sns.barplot(x='id' , y='distance' ,  data=max_distance.head(10))

We can see from the above plot that customer id'id2306955' has the highest distance for trip 

In [ ]:
df.dtypes

In [ ]:
#Changing the data type of variables 'pickup_datetime' and 'dropoff_datetime' to datetime
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])
df['dropoff_datetime'] = pd.to_datetime(df['dropoff_datetime'])

In [ ]:
df["year"] = df["pickup_datetime"].apply(lambda x: x.year)
df["pickup_month"] = df["pickup_datetime"].apply(lambda x: x.month)
df["dropoff_month"] = df["dropoff_datetime"].apply(lambda x: x.month)
df["pickup_day"] = df["pickup_datetime"].apply(lambda x: x.weekday())
df["dropoff_day"] = df["dropoff_datetime"].apply(lambda x: x.weekday())
df["dropoff_time_hour"] = df["dropoff_datetime"].apply(lambda x: x.hour)
df["dropoff_time_min"] = df["dropoff_datetime"].apply(lambda x: x.minute)
df["pick_up_time_hour"] = df["pickup_datetime"].apply(lambda x: x.hour)
df["pick_up_time_min"] = df["pickup_datetime"].apply(lambda x: x.minute)
df.head(3)

## Number of pickups done on each day of the week

In [ ]:
figure, ax = plt.subplots(nrows=2, ncols = 1, figsize = (10,10))
sns.countplot(x = 'pickup_day', data = df, ax = ax[0])
ax[0].set_title('no of pickups done on each day of the week')

sns.countplot(x = 'dropoff_day', data = df, ax=ax[1])
ax[1].set_title('no of dropoffs done on each day of the week')

plt.tight_layout()

In the above plot day-'0' represents 'Monday' and is considerd as the beginning of the week. day-'6' represents Sunday,the end of the week.From the above plot we can see that the number of pickups and dropoffs are maximum for 'Friday' which is represented as day-'4'. 

## Now,let us analyse the trip duration for a  week

duration=df.groupby("pickup_day")['trip_duration'].mean().sort_values(by='trip_duration' , ascending=False)
duration

In [ ]:
duration=df.groupby("pickup_day")['trip_duration'].mean().reset_index().sort_values(by='trip_duration' , ascending=False)
duration

In [ ]:
#Graphical Representation
sns.pointplot(x='pickup_day' , y='trip_duration' , data=duration)
plt.ylabel('Trip Duration')
plt.xlabel('Weekday')
plt.title('Trip Duration per WeekDay')
plt.show()

From the above plot we can infer that the duration of trip is maximum on Thursday represented as day-'3'.

Now let us see the customer who have highest trip duration 

In [ ]:
customer_distance=df.groupby("id")["trip_duration"].mean().reset_index().sort_values(by="trip_duration" , ascending=False)
customer_distance



Now we will find out the number of customers who travel within 30 mins,customers who travel between 30 mins to 1hr and customers who travel from 1hr to 2hr and customers who travel above 2hrs and more.

In [ ]:
#Dividing the dataset on the basis of cutomer travelling time.
#Determining the count of customers who travel within 30 mins
df_travel_time_30=df[df["trip_duration"]<=1800]
df_travel_time_30.head()


In [ ]:
print(len(df_travel_time_30))

We can see from the above data that the count of customer travelling within 30 mins is really very large which is around 1345526

In [ ]:
#Determinig the count of customers who travel from 30 mins to 1hour.
df_travel_time_60=df[(df["trip_duration"]<=3600)& (df["trip_duration"]>1800)]
df_travel_time_60.head()

In [ ]:
print(len(df_travel_time_60))

From the above analysis we can see that the number of customers who travel between 30 mins to 1 hour is 100801.It is pretty less compared to number of customer who travel within 30 mins. 

In [ ]:
#Determiming the count of customers who travel between 1 hours to 2 hours.
df_travel_time_2hr=df[(df["trip_duration"]>3600) & (df["trip_duration"]<=(3600*2))]
df_travel_time_2hr.head()

In [ ]:
print(len(df_travel_time_2hr))

We can see that the number of customers who travel between 1 hour to 2hour is 10064.

In [ ]:
#Determiming the count of customers who travel for more than 2 hours.
df_travel_time=df[df["trip_duration"]>(3600*2)]
df_travel_time.head()

In [ ]:
print(len(df_travel_time))

From the above analysis we can see that the number of customers having travel time of above 2 hours is 2253 which is pretty less compared to the number of customers travelling within 30 mins.

According to a perception we will get a clear view of customers and their tavel 
duration by graphical representaions.However for analysing the customers trip duration it is necessary to deep dive into the data set to unearth many insights.So I have created a new column called "Distance" using the haversine formula.I have analysed the customers with highest  distance and highest trip_duration.
I have broken down the "pickup_datetime" column,"droppoff_datetime" column into 8 columns which are:-
"pickup_month" , "pick_up_time_hr" , "pick_up_time_min" , "pick_up_day",
"dropoff_month" , "dropoff_time_hr" , "dropoff_time_min" , "dropoff_day"
Following are the insights that were analysed:

*   The maximum cabs were booked on "Friday" represented as day-4

*   The highest duration of trip was on "Thursday" reprsented as day-3

*   The highest duration was done by cutomers whose booking id was 'id0053347'.
    It is found out that the trip duration for this trip is 979.52 hours.

*   The highest distance was travelled by the customers whose booking id is
    found to be 'id2306955'.The distance travelled is approximately  1240.91

*   We can see from the  data analysis that the count of customer      
    travelling within 30 mins is really very large which is around '1345526'   

*   From the above analysis it is inferred that the number of customers who 
    travelbetween 30 mins to 1 hour is 100801.It is pretty less compared to   
    number of customer who travel within 30 mins.

*   We can see that the number of customers who travel between 1 hour to 2hour 
    is 10064.It has become less when compared to the number of customer 
    travelling within 30 mins and number of customer travelling between 30 mins
    to 1 hour

*   From the above analysis we can see that the number of customers having  
    travel time of above 2 hours is 2253 which is very less compared to the  
    number of customers travelling within 2hrs.















## ***4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables***

#### Chart - 1 - Horizontal Bar plot on Dependant Variable i.e., Trip Duration (Univariate)

In [ ]:
# Chart - 1 visualization code
sns.set_style('white')
plt.rcParams['font.size'] = 10
plt.rcParams['figure.figsize'] = (14,5)

In [ ]:
df.trip_duration.groupby(pd.cut(df.trip_duration, np.arange(0,9000,600))).count().plot(kind='barh')
plt.xlabel('Trip Counts')
plt.ylabel('Trip Duration in seconds ')
plt.show()

##### 1. Why did you pick the specific chart?